### PR TITLE
Add more demo objects to the demo suite

### PR DESCRIPTION
**Pull request by uittenbroekrobbert**
# Description

Create a demo environment with 4 status columns, 3 example tasks and a user. The status names have been updated to be picked up by the translation module. This PR also renames the statuses created in a previous version so translation works.

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [X]  I have tested the changes locally and verified that they work as expected.
- [X]  I have followed the project's coding conventions and style guidelines.
- [X]  I have rebased my branch onto the latest commit of the main branch.
- [X]  I have squashed or reorganized my commits into logical units.
- [X]  I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

### CODE REVIEW

File: tad/core/db.py [hunk 1]
```diff
@@ -44,20 +44,40 @@ def init_db():
...
+        user = User(name="Robbert", avatar=None)
+        session.add(user)
 
-            status = session.exec(select(Status).where(Status.name == "Todo")).first()
-            if not status:
-                status = Status(name="Todo", sort_order=1)
-                session.add(status)
 
-            task = session.exec(select(Task).where(Task.title == "First task")).first()
+def _add_demo_tasks(session: Session, todo_status: Status | None) -> None:
```
Comment (berrydenhartog) [inline tad/core/db.py:61]:
```suggestion
def _add_demo_tasks(session: Session, status: Status | None) -> None:
```

File: tad/core/db.py [hunk 2]
```diff
@@ -44,20 +44,40 @@ def init_db():
...
+    if todo_status is not None:
```
Comment (berrydenhartog) [inline tad/core/db.py:62]:
```suggestion
    if status is not None:
```

File: tad/core/db.py [hunk 3]
```diff
@@ -44,20 +44,40 @@ def init_db():
...
+        for index in range(1, 3):
```
Comment (berrydenhartog) [inline tad/core/db.py:63]: i would add this logic outside of this function

**Review by berrydenhartog — COMMENTED** [review]
Looks good. I would make the functions generic

**Comment by uittenbroekrobbert** [comment]
> Looks good. I would make the functions generic

Hm, I'm not sure what you mean or how that helps. Can you elaborate?

File: tests/core/test_db.py [hunk 4]
```diff
@@ -32,14 +44,27 @@ def test_init_database_none(patch_settings: Settings):
 
     init_db()
 
-    expected = [
-        (select(User).where(User.name == "Robbert"),),
-        (select(Status).where(Status.name == "Todo"),),
-        (select(Task).where(Task.title == "First task"),),
-    ]
+    for i, call_args in enumerate(Session.exec.call_args_list):
```
Comment (berrydenhartog) [inline tests/core/test_db.py:47]: it probably better to use the pytest.parametrize option to handle multiple test cases with the same logic. 

See: https://docs.pytest.org/en/7.0.x/how-to/parametrize.html

**Comment by berrydenhartog** [comment]
> > Looks good. I would make the functions generic
> 
> Hm, I'm not sure what you mean or how that helps. Can you elaborate?

I mean that i would take hardcodes arrays outside of the function. example:

```
def add_demo_statuses(session: Session) -> None:
    for index, status_name in enumerate(["todo", "review", "in_progress", "done"]):
        status = session.exec(select(Status).where(Status.name == status_name)).first()
        if not status:
            status = Status(name=status_name, sort_order=index + 1)
            session.add(status)
            session.commit()
            session.commit()
```

i would rewrite to:
```
def add_demo_statuses(session: Session, status_name: String, index: int) -> None:
        status = session.exec(select(Status).where(Status.name == status_name)).first()
        if not status:
            status = Status(name=status_name, sort_order=index)
            session.add(status)
            session.commit()
            session.commit()
```

This would allow more flexibility in what demo mode is.

File: tests/core/test_db.py (hunk 4 above)
Comment (uittenbroekrobbert) [reply tests/core/test_db.py:41]: I think I did that now. Can you check @berrydenhartog ?

File: tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py [hunk 5]
```diff
@@ -21,7 +21,7 @@
 
 def upgrade() -> None:
     # ### commands auto generated by Alembic - please adjust! ###
-    status = op.create_table(
+    op.create_table(
```
Comment (berrydenhartog) [inline tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py:24]: you should not edit migrations. You will need to create a new migration that undoes the current migration

File: tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py [hunk 6]
```diff
@@ -56,17 +56,6 @@ def upgrade() -> None:
     )
     # ### end Alembic commands ###
 
-    # ### custom commands ###
-    op.bulk_insert(
```
Comment (berrydenhartog) [inline tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py:60]: Why is this removed?

File: tad/core/db.py [hunk 7]
```diff
@@ -43,21 +56,59 @@ def init_db():
...
-                session.add(task)
-            session.commit()
+            remove_old_demo_objects(session)
+            add_demo_users(session, ["default user"])
+            demo_statuses = add_demo_statuses(session, ["todo", "review", "in_progress", "done"])
+            add_demo_tasks(session, demo_statuses[0], 3)
     logger.info("Finished initializing database")
+
+
+def truncate_tables(session: Session) -> None:
```
Comment (anneschuth) [inline tad/core/db.py:69]: This is rather scary code to have, can we guard it with something like:

```python
if get_settings().ENVIRONMENT == "prod":
    return
```

Comment (uittenbroekrobbert) [reply tad/core/db.py:69]: I think no one needed this besides me and there are others ways for what I need. I will delete this function as it undoubtedly will create an unintended database wipe in the future.

File: tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py (hunk 5 above)
Comment (uittenbroekrobbert) [reply tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py:24]: My bad. I will create a new migration.

File: tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py (hunk 6 above)
Comment (uittenbroekrobbert) [reply tad/migrations/versions/b62dbd9468e4_create_status_user_and_task_table.py:60]: My bad, I will make it a migration to rename the exisitng columns so the translations will work.

File: tad/core/db.py (hunk 7 above)
Comment (berrydenhartog) [reply tad/core/db.py:69]: It would be better to guard it in the tad/core/config as a model_validator:

```
    @model_validator(mode="after")
    def _enforce_database_cleanup_rules(self: SelfSettings) -> SelfSettings:
        if self.ENVIRONMENT == "production" and self.TRUNCATE_TABLES:
            raise SettingsError("TRUNCATE_TABLES=True is not supported in production")
        return self
```

Comment (uittenbroekrobbert) [reply tad/core/db.py]: Dat is wel heel cool. Die code zal ik zeker onthouden. Voor nu heb ik de truncate er uit gelaten. Het is niet perse nodig en daarmee voorkomen we dat het uberhaubt fout kan gaan ;)

File: tests/core/test_db.py [hunk 8]
```diff
@@ -21,47 +35,76 @@ def test_check_database():
...
-        (select(User).where(User.name == "Robbert"),),
-        (select(Status).where(Status.name == "Todo"),),
-        (select(Task).where(Task.title == "First task"),),
-    ]
-
-    for i, call_args in enumerate(Session.exec.call_args_list):
-        assert str(expected[i][0]) == str(call_args.args[0])
-
-    Session.exec = org_exec
+    if patch_settings.ENVIRONMENT == "demo":
```
Comment (berrydenhartog) [inline tests/core/test_db.py:48]: Why do we now check the logs instead of the actual functions called? I prefer the actual functions because then you check what is actually happening instead of the logs which is not the main logic of the function.

File: tests/core/test_db.py [hunk 9]
```diff
@@ -21,47 +35,76 @@ def test_check_database():
...
+    assert [record.message for record in caplog.records if "Finished initializing database" in record.message]
+
+
+def test_remove_old_demo_objects(db: DatabaseTestUtils, monkeypatch: pytest.MonkeyPatch):
+    user = User(name="Robbert", avatar=None)
+    status = Status(name="Todo", sort_order=1)
+    task = Task(title="First task", description="This is the first task", sort_order=1, status_id=status.id)
+
+    db.given([user, status, task])
+    session_delete_counter = count_calls(monkeypatch, db.get_session(), db.get_session().delete)
```
Comment (berrydenhartog) [inline tests/core/test_db.py:61]: Why do we not use magicmock for the delete function here if you just want to know the amount of calls? 

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_count

File: tad/core/db.py [hunk 10]
```diff
@@ -44,20 +57,48 @@ def init_db():
...
-            if not status:
-                status = Status(name="Todo", sort_order=1)
-                session.add(status)
-
-            task = session.exec(select(Task).where(Task.title == "First task")).first()
-            if not task:
-                task = Task(title="First task", description="This is the first task", sort_order=1, status_id=status.id)
-                session.add(task)
-            session.commit()
+            remove_old_demo_objects(session)
```
[the remark this answers is not in the document]
Comment (uittenbroekrobbert) [reply tad/core/db.py:60]: The alembic database script creates 4 statuses. These are not 'demo related' I think. So truncating would delete those. The 'remove old objects' deletes only the objects created by the previous demo function. So I think this would be the best way.

File: tests/core/test_db.py (hunk 8 above)
Comment (uittenbroekrobbert) [reply tests/core/test_db.py:48]: Valid point. I did not know a better way to check if a function had been called, unless using a Mock which I prefer not to use because then I don't know if the logic works (only that the function has been called). I did however find a 'count calls' decorater, I will see if I can use that.

File: tests/core/test_db.py (hunk 9 above)
Comment (uittenbroekrobbert) [reply tests/core/test_db.py:61]: I want to know the logic works, so no exception is thrown f.e. and to make sure t is called as often as expected I use the counter.

File: tests/core/test_db.py (hunk 8 above)
Comment (uittenbroekrobbert) [reply tests/core/test_db.py:48]: Maybe you can help @berrydenhartog as I have been using a wrapper around a function to be able to count calls, but in this case need to change an existing function and I can't get that to work.

File: tad/core/db.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (berrydenhartog) [reply tad/core/db.py:60]: I agree

**Review by berrydenhartog — APPROVED** [review]
LGTM
